pkg/node/client.go

```diff
@@ -0,0 +1,165 @@
+package node
+
+import (
+	"context"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	cs "github.com/kubernetes-sigs/container-object-storage-interface-api/clientset/typed/objectstorage.k8s.io/v1alpha1"
+)
+
+type NodeClient struct {
+	cosiClient *cs.ObjectstorageV1alpha1Client
+	kubeClient kubernetes.Interface
+}
+
+func NewClient() (*NodeClient, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	client := cs.NewForConfigOrDie(config)
+	kube := kubernetes.NewForConfigOrDie(config)
+	return &NodeClient{
+		cosiClient: client,
+		kubeClient: kube,
+	}, nil
+}
+
+const (
+	barNameKey      = "bar-name"
+	barNamespaceKey = "bar-namespace"
+)
+
+func parseValue(key string, ctx map[string]string) (string, error) {
+	value, ok := ctx[key]
+	if !ok {
+		return "", fmt.Errorf("required volume context key unset: %v", key)
+	}
+	klog.Infof("got value: %v", value)
+	return value, nil
+}
+
+func parseVolumeContext(volCtx map[string]string) (name, ns string, err error) {
+	klog.Info("parsing bucketAccessRequest namespace/name from volume context")
+	name, err = parseValue(barNameKey, volCtx)
+	if err != nil {
+		return
+	}
+	ns, err = parseValue(barNamespaceKey, volCtx)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (n *NodeClient) getBAR(ctx context.Context, barName, barNs string) (*v1alpha1.BucketAccessRequest, error) {
+	klog.Infof("getting bucketAccessRequest %q", fmt.Sprintf("%s/%s", barNs, barName))
+	bar, err := n.cosiClient.BucketAccessRequests(barNs).Get(ctx, barName, metav1.GetOptions{})
+	if err != nil {
+		return nil, logErr(getError("bucketAccessRequest", fmt.Sprintf("%s/%s", barNs, barName), err))
+	}
+	// TODO: need to enable validation after resolving status issue - Krish
+	//if bar == nil || !bar.Status.AccessGranted {
+	//	return nil, logErr(fmt.Errorf("BucketAccessRequest is nil or BucketAccessRequest does not grant access %q", fmt.Sprintf("%s/%s", barNs, barName)))
```

Comment:
maybe for better log readability, split this compound error into 2 separate error checks.

Reply:
Done! I also added some more validation. One unrelated question, I use the BucketInstanceName from the BR, but the BA also has a BucketInstanceName. These should always be the same as far as I know, so it shouldn't matter which one is used, correct?

Reply:
BA.spec.bucketInstanceName should always match BR.spec.bucketInstanceName.
If COSI evolves where a BR is not needed for brownfield then the cosi-node-adapter could not rely on a BR, so maybe it's better to use the BA for this field.